---
title: "extraChIPs: Differential Signal Using Fixed-Width Windows"
author:
- name: Stevie Pederson
  affiliation: 
  - Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide, Australia
  - Dame Roma Mitchell Cancer Researc Laboratories, University of Adelaide
  - John Curtin School of Medical Research, Australian National University
  email: stephen.pederson.au@gmail.com
package: extraChIPs
bibliography: '`r system.file("references.bib", package = "extraChIPs")`'
output:
  github_document
vignette: >
  %\VignetteIndexEntry{Differential Signal Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(
  message = FALSE, warning = FALSE, fig.height = 6, fig.width = 8
)
```

# Introduction

The [GRAVI](https://github.com/smped/GRAVI) workflow, for which this package 
is designed, uses sliding windows for differential signal analysis.
However, the use of fixed-width windows, as is common under DiffBind-style
[@DiffBind2012] approaches is also possible with `extraChIPs`.
This vignette focusses on using conventional peak calls and fixed-width 
approaches to replicate and extend these approaches.

The majority of examples below use heavily reduced datasets to provide
general guidance on using the functions.
Some results may appear trivial as a result, but will hopefully prove far more
useful in a true experimental context.
All data, along with this vignette are available 
[here](https://github.com/smped/extraChIPs_vignette).
Please place all contents of the data directory in a directory named data in 
your own working directory.

# Setup 

## Installation

In order to use the package `extraChIPs` and follow this vignette, we recommend 
using the package `BiocManager` hosted on CRAN.
Once this is installed, the additional packages required for this vignette 
(`tidyverse`, `Rsamtools`, `csaw`, `BiocParallel` and `rtracklayer`) can also 
be installed.

```{r install, eval = FALSE}
if (!"BiocManager" %in% rownames(installed.packages()))
  install.packages("BiocManager")
pkg <- c(
  "tidyverse", "Rsamtools", "csaw", "BiocParallel", "rtracklayer", "edgeR", 
  "patchwork", "extraChIPs", "plyranges", "scales", "here", "quantro"
)
BiocManager::install(pkg, update = FALSE)
```

Once these packages are installed, we can load them easily

```{r load-packages}
library(tidyverse)
library(Rsamtools)
library(csaw)
library(BiocParallel)
library(rtracklayer)
library(edgeR)
library(patchwork)
library(extraChIPs)
library(plyranges)
library(scales)
library(glue)
library(ggrepel)
library(here)
library(quantro)
theme_set(theme_bw())
```


## Data

All data for this vignette is expected to be in a sub-directory of the working 
directory named "data", and all paths will be predicated on this.
Please ensure you have all data in this location, obtained from 
[here](https://github.com/smped/extraChIPs_vignette).

The data itself is ChIP-Seq data targeting the Estrogen Receptor (ER), and is 
taken from the cell-line ZR-75-1 cell-line using data from the BioProject , 
Pre-processing was performed using the 
[`prepareChIPs`](https://github.com/smped/prepareChIPs) workflow, written in 
snakemake [@Molder2021-mo] and all code is available at 
https://github.com/smped/PRJNA509779.
ER binding was assessed under Vehicle (E2) and DHT-stimulated (E2DHT) conditions.
Using GRCh37 as the reference genome, a subset of regions found on chromosome 10
are included in this dataset for simplicity.

First we'll load our sample data then define our two treatment groups.
Defining a consistent colour palette for all plots is also a good habit to develop.

```{r load-samples}
samples <- here("data", "PRJNA509779.tsv") %>% 
  read_tsv() %>% 
  dplyr::filter(target == "ER") %>% 
  mutate(treatment = factor(treatment, levels = c("E2", "E2DHT")))
samples
accessions <- samples %>% 
  split(f = .$treatment) %>% 
  lapply(pull, "accession")
treat_levels <- levels(samples$treatment)
treat_colours <- setNames(c("steelblue", "red3"), treat_levels)
```

We'll eventually be loading counts for differential signal analysis from a set 
of BamFiles, so first we'll create a `BamFileList` with all of these files.
This also enables creation of a `Seqinfo` object based on the actual reference 
genome to which the reads were aligned during data preparation.
`Seqinfo` objects are the foundation of working with GRanges, so defining an 
object at the start of a workflow is good practice.

```{r bfl}
bfl <- here("data", "ER", glue("{samples$accession}.bam")) %>% 
  BamFileList() %>% 
  setNames(str_remove_all(names(.), ".bam"))
file.exists(path(bfl))
sq <- seqinfo(bfl)
isCircular(sq) <- rep(FALSE, length(seqlevels(sq)))
genome(sq) <- "GRCh37"
```

Another key preparatory step for working with peaks is to define a set of 
regions as either blacklisted or grey-listed regions.
The former are known problematic regions based on each genome, with data freely 
available from https://github.com/Boyle-Lab/Blacklist/tree/master/lists, whilst 
grey-listed regions are defined from potentially problematic regions as detected
within the input sample.
For our samples code for this is included in the previously provided repository 
(https://github.com/smped/PRJNA509779).

```{r greylist}
greylist <- import.bed(here("data/chr10_greylist.bed"), seqinfo = sq)
blacklist <- import.bed( here("data/chr10_blacklist.bed"), seqinfo = sq)
omit_ranges <- c(greylist, blacklist)
```


# Working With Peaks

The provided dataset includes six files produced by `macs2 callpeak` 
[@Zhang2008-ms] in the `narrowPeak` format, and these are able to be easily 
parsed using `extraChIPs`.
We'll immediately pass our black & grey-listed regions to our parsing function 
so we can exclude these regions right from the start

```{r load-peaks}
peaks <- here("data", "ER", glue("{samples$accession}_peaks.narrowPeak")) %>% 
  importPeaks(seqinfo = sq, blacklist = omit_ranges)
```

This will import the peaks from all files as a single `GRangesList` object, 
adding the file-name to each element by default.
We can easily modify these names if we so wish.

```{r change-names}
names(peaks) <- str_remove_all(names(peaks), "_peaks.narrowPeak")
```

Once loaded, we can easily check how similar our replicates are using 
`plotOverlaps()`.
When three or more sets of peaks are contained in the `GRangesList`, an UpSet 
plot will be drawn by default.

```{r plot-overlaps, fig.height=6, fig.cap = "*UpSet plot showing overlapping peaks across all replicates*"}
plotOverlaps(
  peaks, min_size = 10, .sort_sets = FALSE, 
  set_col = treat_colours[as.character(samples$treatment)]
)
```

Optionally, specifying a column and a suitable function will produce an 
additional panel summarising that value.
In the following, we'll show the maximum score obtained, highlighting that for 
peaks identified in only one or two replicates, the overall signal intensity is
generally lower, even in the sample with the strongest signal.

```{r plot-overlaps-score, fig.height = 7, fig.cap = "*UpSet plot showing overlapping peaks across all replicates, with the maximum score across all replicates shown in the upper panel.*"}
plotOverlaps(
  peaks, min_size = 10, .sort_sets = FALSE, var = "score", f = "max",
   set_col = treat_colours[as.character(samples$treatment)]
)
```

A common task at this point may be to define consensus peaks within each 
treatment group, by retaining only the peaks found in 2 of the 3 replicates 
`(p = 2/3)`.
The default approach is to take the union of all ranges, with the returned 
object containing logical values for each sample, as well as the number of 
samples where an overlapping peak was found.

If we wish to retain any of the original columns, such as the `macs2 callpeak` 
score, we can simply pass the column names to `makeConsensus()`

```{r consensus-with-score}
consensus_e2 <- makeConsensus(peaks[accessions$E2], p = 2/3, var = "score")
consensus_e2dht <- makeConsensus(peaks[accessions$E2DHT], p = 2/3, var = "score")
```

Alternatively, we could find the centre of the peaks as part of this process, 
by averaging across the estimated peak centres for each sample.
This is a very common step for ChIP-Seq data where the target is a 
*transcription factor*, and also forms a key step in the DiffBind workflow.

In the following code chunk, we first find the centre for each sample using the 
information provided by `macs2`, before retaining this column when calling 
`makeConsensus()`.
This will return each of the individual centre-position estimates as a list for 
each merged range, and using `vapply()` we then take the mean position as our 
estimate for the combined peak centre.

```{r consensus-with-centre}
consensus_e2 <- peaks[accessions$E2] %>% 
  endoapply(mutate, centre = start + peak) %>% 
  makeConsensus(p = 2/3, var = "centre") %>% 
  mutate(centre = vapply(centre, mean, numeric(1)))
consensus_e2
consensus_e2dht <- peaks[accessions$E2DHT] %>% 
  endoapply(mutate, centre = start + peak) %>% 
  makeConsensus(p = 2/3, var = "centre") %>% 
  mutate(centre = vapply(centre, mean, numeric(1)))
```

We can also inspect these using `plotOverlaps()` provided we use a `GRangesList`
for the input.
Now that we only have two elements (one for each treatment) a VennDiagram will 
be generated instead of an UpSet plot.

```{r venn-consensus-overlap, fig.width=6, fig.cap = "*Overlap between consensus peaks identified in a treatment-specific manner*"}
GRangesList(E2 = granges(consensus_e2), E2DHT = granges(consensus_e2dht)) %>% 
  plotOverlaps(set_col = treat_colours[treat_levels])
```

We can now go one step further and define the set of peaks found in either 
treatment.
Given we're being inclusive here, we can leave p = 0 so any peak found in 
*either treatment* is included.

```{r union-peaks}
union_peaks <- GRangesList(
  E2 = select(consensus_e2, centre), 
  E2DHT = select(consensus_e2dht, centre)
) %>% 
  makeConsensus(var = c("centre")) %>% 
  mutate(
    centre = vapply(centre, mean, numeric(1)) %>% round(0)
  ) 
```

Now we have a set of peaks, found in at least 2/3 of samples from either 
condition, with estimates of each peak's centre.
The next step would be to set all peaks as the same width based on the centre 
position, with a common width being 500bp.

In the following we'll perform multiple operations in a single call mutate, so 
let's make sure we know what's happening.

1. `glue("{seqnames}:{centre}:{strand}")` uses `glue` syntax to parse the 
seqnames, centre position and strand information as a character-like vector
with a width of only 1, and using the estimated centre as the Range.
2. We then coerce this to a `GRanges` object, before resizing to the desired 
width.
3. We also add the original (un-centred) range as an additional column, 
retaining the `GRanges` structure, but discarding anything in the `mcols()` 
element, then
4. Using `colToRanges()`, we take the centred ranges and place them as the core
set of GRanges for this object.

This gives a GRanges object with all original information, but with centred 
peaks of a fixed width.

```{r centred-peaks}
w <- 500
centred_peaks <- union_peaks %>% 
  mutate(
    centre = glue("{seqnames}:{centre}:{strand}") %>% 
      GRanges(seqinfo = sq) %>% 
      resize(width = w),
    union_peak = granges(.)
  ) %>% 
  colToRanges("centre")
```

# Counting Reads

Now we have our centred, fixed-width peaks, we can count reads using 
`csaw::regionCounts()` [@csaw2016].
We know our fragment length is about 200bp, so we can pass this to the function 
for a slightly more sophisticated approach to counting.

```{r count-reads}
se <- regionCounts(bfl, centred_peaks, ext = 200)
se
```

The `colData()` element of the returned object as the columns 
`r pander::pander(colnames(colData(se)))`, which are all informative and can be 
supplemented with our `samples` data frame.
In the following, we'll 1) coerce to a `tibble`, 2) `left_join()` the `samples`
object, 3) add the accession as the sample column, 4) set the accession back as 
the rownames, then 5) coerce back to the required `DataFrame()` structure.

```{r tidy-coldata}
colData(se) <- colData(se) %>% 
  as_tibble(rownames = "accession") %>% 
  left_join(samples) %>% 
  mutate(sample = accession) %>% 
  as.data.frame() %>% 
  column_to_rownames("accession") %>% 
  DataFrame()
colData(se)
```

For QC and visualisation, we can add an additional `logCPM` assay to our object 
as well.

```{r add-logcpm}
assay(se, "logCPM") <- cpm(assay(se, "counts"), lib.size = se$totals, log = TRUE)
```

First we might like to check our distribution of counts

```{r plot-assay-densities, fig.cap = "*Count densities for all samples, using the log+1 transformation*"}
plotAssayDensities(se, assay = "counts", colour = "treat", trans = "log1p") +
  scale_colour_manual(values = treat_colours)
```

An RLE plot also can be informative

```{r plot-rle, fig.cap = "*RLE plot using logCPM values*"}
plotAssayRle(se, assay = "logCPM", fill = "treat") +
  scale_fill_manual(values = treat_colours)
```

Finally a PCA plot can also provide insight as to where the variability in the data lies.

```{r plot-pca, fig.cap = "*PCA plot using logCPM values and showing that replicate variability is larger than varibility between treatment groups.*"}
plotAssayPCA(se, assay = "logCPM", colour = "treat", label = "sample") +
  scale_colour_manual(values = treat_colours)
```

# Differential Signal Analysis

## Statistical Testing


In order to perform Differential Signal Analysis, we simply need to define a 
model matrix, as for conventional analysis using `edgeR` or `limma`.
We can then pass this, along with our fixed-width counts to `fitAssayDiff()`.
By default normalisation will be *library-size* normalisation, as is a common 
default strategy for ChIP-Seq data.
In contrast to sliding window approaches, these results represent our final 
results and there is no need for merging windows.

```{r fit-library-size}
X <- model.matrix(~treatment, data = colData(se))
ls_res <- fitAssayDiff(se, design = X, asRanges = TRUE)
sum(ls_res$FDR < 0.05)
```

TMM normalisation [@Robinson2010-qp] is another common strategy, which relies 
on the data from all treatment groups being drawn from the same distributions.
We can formally test this using the package `quantro` [@Hicks2015-ee] , which 
produces p-values for 1) H~0~: Group medians are drawn from the same 
distribution, and 2) H~0~: Group-specific distributions are the same.

```{r qtest}
set.seed(100)
qtest <- assay(se, "counts") %>% 
  quantro(groupFactor = se$treatment, B = 1e3)
qtest
```

Here, both p-values are >0.05, so in conjunction with out visual inspection 
earlier, we can confidently apply TMM normalisation.
To apply this, we simply specify the argument `norm = "TMM"` when we call 
`fitAssayDiff()`.
In the analysis below, we've also specified a fold-change threshold 
`(fc = 1.2)`, below which, changes in signal are considered to not be of 
interest [@McCarthy2009-qf].
This threshold is incorporated into the testing so there is no requirement for 
*post-hoc* filtering based on a threshold.

```{r fit-tmm}
tmm_res <- fitAssayDiff(se, design = X, norm = "TMM", asRanges = TRUE, fc = 1.2)
sum(tmm_res$FDR < 0.05)
```


An MA-plot is a common way of inspecting results and in the following we use 
the original 'union_peak' in our labelling of points.
This serves as a reminder that the fixed-width windows are in fact *a proxy* for 
the entire region for which we have confidently detected ChIP signal, and that 
these windows are truly the regions of interest.

```{r plot-ma, fig.cap = "*MA-plot after fitting using TMM normalisation and applying a fold-change threshold during testing. Points are labelled using the original windows obtained when merging replicats and treatment groups.*"}
tmm_res %>% 
  as_tibble() %>% 
  mutate(`FDR < 0.05` = FDR < 0.05) %>% 
  ggplot(aes(logCPM, logFC)) +
  geom_point(aes(colour = `FDR < 0.05`)) +
  geom_smooth(se = FALSE) +
  geom_label_repel(
    aes(label = union_peak), colour = "red",
    data = . %>% dplyr::filter(FDR < 0.05)
  ) +
  scale_colour_manual(values = c("black", "red"))
```

## Mapping to Genes

Whilst knowledge of which regions are showing differential signal, the 
fundamental question we are usually asking is about the downstream regulatory 
consequences, such as the target gene.
Before we can map peaks to genes, we'll need to define our genes.
In the following, we'll use the provided Gencode gene mappings at the gene, 
transcript and exon level.

```{r gencode}
gencode <- here("data/gencode.v43lift37.chr10.annotation.gtf.gz") %>% 
  import.gff() %>% 
  filter_by_overlaps(GRanges("chr10:42354900-100000000")) %>% 
  split(.$type)
seqlevels(gencode) <- seqlevels(sq)
seqinfo(gencode) <- sq
```

Mapping to genes using `mapByFeature()` uses additional annotations, such as 
whether the peak overlaps a promoter, enhancer or long-range interaction.
Here we'll just use promoters, so let's create a set of promoters from our 
transcript-level information, ensuring we incorporate all possible promoters 
within a gene, and merging any overlapping ranges using `reduceMC()`

```{r promoters}
promoters <- gencode$transcript %>% 
    select(gene_id, ends_with("name")) %>% 
    promoters(upstream = 2500, downstream = 500) %>% 
    reduceMC(simplify = FALSE)
promoters
```

Now we'll pass these to `mapByFeature()`, but first, we'll place the original 
'union_peak' back as the core of the GRanges object.
This will retain all the results from testing, but ensures the correct region is
mapped to genes.

```{r tmm-mapped}
tmm_mapped_res <- tmm_res %>% 
  colToRanges("union_peak") %>% 
  mapByFeature(genes = gencode$gene, prom = promoters) %>% 
  mutate(
    status = case_when(
      FDR >= .05 ~ "Unchanged",
      logFC > 0 ~ "Increased",
      logFC < 0 ~ "Decreased"
    )
  )
arrange(tmm_mapped_res, PValue)
```


# Inspection of Results

## Profile Heatmaps

When analysing a transcription factor, checking the binding profile across our 
treatment groups can be informative, and is often performed using 'Profile 
Heatmaps' where coverage is smoothed within bins surrounding our peak centre.

The function `getProfileData()` takes a set of ranges and a BigWigFileList, and 
performs the smoothing, which is then passed to the function 
`plotProfileHeatmap()`.
Let's first check the basic coverage using the SPMR-scaled values as produced 
by `macs2`.
These are equivalent to CPM values and can be log-transformed as required.

```{r cov-bw}
cov_bw <- here::here(
  "data", "ER", glue("{levels(samples$treatment)}_cov_chr10.bw")
) %>% 
  BigWigFileList() %>% 
  setNames(treat_levels)
```


The following shows the three steps of 1) defining the ranges, 2) obtaining the 
smoothed binding profiles, and 3) drawing the heatmap.
Note that we can facet the heatmaps by selecting the 'status' column to separate 
any Increased or Decreased regions.
By default, this will also draw the smoothed lines in the top panel using 
different colours.

(Note the SPMR values were calculated using the complete genome and will differ 
significantly from those obtained during model fitting above, as this will only 
use library sizes for our highly restricted subset of the genome.)

```{r cov-profile-heatmaps, fig.cap = "*Profile Heatmap showing read coverage for all sites considered as showing evidence of differential signal. The sites are annotated by status using facets along the y-axis*"}
sig_ranges <- filter(tmm_mapped_res, FDR < 0.05)
pd <- getProfileData(cov_bw, sig_ranges)
plotProfileHeatmap(pd, "profile_data", facetY = "status") +
  scale_fill_gradient(low = "white", high = "red") +
  labs(fill = "logCPM", colour = "Status")
```

As an alternative to CPM-like coverage values, we can use fold-enrichment over 
the input sample(s), as is also produced by `macs2 bdgcmp`.
This data isn't generally visualised using log-transformation so we'll set 
`log = FALSE` in our call to `getProfileData()`

```{r fe-profile-heatmaps, fig.cap = "*Profile Heatmap showing fold-enrichment over input samples for all sites consiered as showing evidence of differential signal.*"}
fe_bw <- here("data", "ER", glue("{treat_levels}_FE_chr10.bw")) %>% 
  BigWigFileList() %>% 
  setNames(treat_levels)
pd_fe <- getProfileData(fe_bw, sig_ranges, log = FALSE) 
pd_fe %>% 
  plotProfileHeatmap("profile_data") +
  scale_fill_gradient(low = "white", high = "red") +
  labs(fill = "Fold\nEnrichment")
```


## Coverage Plots

As well as showing summarised values across all sites, we may wish to check the 
binding patterns in relation to other genomic features.
`plotHFGC()` enables easy visualisation of a region incorporating HiC (H), 
Features (F), Genes (G) and Coverage (C), with considerable flexibility in how 
many tracks are able to be shown.
`plotHFGC()` relies on the infrastructure provided by `Gviz` [@gviz] so some 
familiarity with this package is helpful, but not essential.

The minimum requirement is a GRanges object and coverage is obviously important, 
so let's choose our most highly-ranked result, which is mapped to the gene 
*PPIF*, and the SPMR-based BigWigFileList.

```{r set-gr}
gr <- arrange(tmm_mapped_res, PValue)[1]
gr
```


```{r plot-hfgc-basic, fig.cap = "*The most simple coverage output for plotHFGC*"}
plotHFGC(gr, coverage = cov_bw)
```

As can be seen from the y-limits, although the shape of the peaks is the same, 
the coverage is significantly different.
We could manually set this to be the same on both tracks

```{r plot-hfgc-ylim, fig.cap = "*The same plot as above, but setting y-axis limits manually*"}
plotHFGC(gr, coverage = cov_bw, ylim = c(0, 5))
```

An alternative may be to plot both treatments on the same track.
By default, if a BigWigFileList is passed to `plotHFGC()` each BigWig file will 
be drawn on a separate track.
To overlap the tracks, we can pass a named list of BigWigFileList objects and a 
corresponding list of colours, using the same structure

```{r plot-hfgc-overlap, fig.cap = "*Coverage tracks can be overlaid to ensure a coparable y-axis*"}
cov_list <- list(ER = cov_bw)
cov_colours <- list(ER = treat_colours)
plotHFGC(gr, coverage = cov_list, linecol = cov_colours)
```

We can add additional coverage tracks using the same strategy.
Given we used H3K27ac data in the sliding windows vignette, let's load H3K27ac 
coverage and add this to our plots.

```{r plot-hfgc-h3k27ac, fig.cap = "*By overlapping tracks, multiple ChIP targets can be shown, here with the addition of H3K27ac coverage*"}
cov_list$H3K27ac <- here::here(
  "data", "H3K27ac", glue("{levels(samples$treatment)}_cov_chr10.bw")
) %>% 
  BigWigFileList() %>% 
  setNames(treat_levels)
cov_colours$H3K27ac <- treat_colours[treat_levels]
plotHFGC(gr, coverage = cov_list, linecol = cov_colours, rotation.title = 90)
```


Now let's zoom out and add some cytogenetic bands for reference

```{r plot-hfgc-cytobands, fig.cap = "*Cytogentic bnds are providing for some genomes with `extraChIPs`. Ranges can also be zoomed in or out as required.*"}
data("grch37.cytobands")
plotHFGC(
  gr, 
  coverage = cov_list, linecol = cov_colours, rotation.title = 90,
  cytobands = grch37.cytobands, zoom = 20
)
```

Next, we might like to add some genes or transcripts.
First we'll use the exon element of our Gencode data to create a suitable 
object for Gviz.

```{r plot-hfgc-genes, fig.cap = "*Genes can be added as a separate track. Here we can see that ER binding is upstream of the gene PPIF, with H3K27ac cverage associated with both the promoter and ER binding, suggesting that our particular range is in an enhancer.*"}
gene_models <- gencode$exon %>% 
  select(
    type, gene = gene_id, exon = exon_id, transcript = transcript_id, 
    symbol = gene_name
  ) %>% 
  keepStandardChromosomes()
plotHFGC(
  gr, 
  genes = gene_models, genecol = "wheat",
  coverage = cov_list, linecol = cov_colours, rotation.title = 90,
  cytobands = grch37.cytobands, zoom = 20
)
```

Given that we've already defined some promoters for our mapping to genes, we 
can add these are features.
For features, we need to pass `plotHFGC()` a named GRangesList, and the names 
are used to provide colours to the features

```{r plot-hfgc-features, fig.cap = "*Any type of genomic feature can be added. Here, possible promoters are shown in yellow*"}
feat_grl <- GRangesList(Promoters = promoters) %>% 
  keepStandardChromosomes()
feat_colours <- list(Promoters = "yellow2")
plotHFGC(
  gr, 
  features = feat_grl, featcol = feat_colours,
  genes = gene_models, genecol = "wheat",
  coverage = cov_list, linecol = cov_colours, rotation.title = 90,
  cytobands = grch37.cytobands, zoom = 20
)
```

We may also wish to add known binding sites for ER as provided by Encode.
The file `esr1_chr10.hg19.bed.gz` was obtained from the UCSC Table Browser, and 
this combines ER (i.e. *ESR1*) binding sites for multiple cell lines.

```{r plot-hfgc-esr1, fig.cap = "*Multiple features can be added to a single track. Here ER sites defined by ENCODE are shown in blue*"}
esr1 <- here("data/ER/esr1_chr10.hg19.bed.gz") %>% 
  import.bed(
    colnames = c("chrom", "start", "end", "name", "score"), seqinfo = sq
  ) %>% 
  keepStandardChromosomes()
feat_grl$ESR1 <- esr1
feat_colours <- list(Promoters = "yellow2", ESR1 = "royalblue")
plotHFGC(
  gr, 
  features = feat_grl, featcol = feat_colours,
  genes = gene_models, genecol = "wheat",
  coverage = cov_list, linecol = cov_colours, rotation.title = 90,
  cytobands = grch37.cytobands, zoom = 20
)
```


We can separate out these features onto separate tracks using a similar 
strategy to our coverage tracks.
If we pass a GRangesList, all features will be drawn on the same track, however, 
if we pass a list of GRangesList objects, each list element will be drawn as a 
separate track.
Our colours object will now need to be specified as a list with the same 
structure.

```{r plothfgc-feat-list, fig.cap = "*Features can also be shown on eparate tracks with informative track labels. Plots can also be shifted and multiple parameters are also able to be customised*"}
feat_list <- list(Promoters = feat_grl["Promoters"], ESR1 = feat_grl["ESR1"])
feat_colours <- list(
  Promoters = c(Promoters = "yellow2"), ESR1 = c(ESR1 = "royalblue")
)
plotHFGC(
  gr, 
  features = feat_list, featcol = feat_colours, featsize = 1.5,
  genes = gene_models, genecol = "wheat",
  coverage = cov_list, linecol = cov_colours, rotation.title = 90,
  cytobands = grch37.cytobands, zoom = 15, shift = 6e3,
  fontsize = 12, cex.axis = 0.8, highlight = rgb(0.7, 0.7, 1)
)
```

Although not demonstrated here, the same principles for multiple gene-level 
tracks can be applied.
This is particularly useful if wishing to plot differentially expressed genes 
on a separate track to those which are undetected or unchanged.

Additional plotting functions which may be useful are also demonstrated in the 
sliding windows vignette.

## References

<div id="refs"></div>
<br>

## Session Info

```{r session-info}
sessionInfo()
```



