---
title: "extraChIPs: Differential Signal Using Fixed-Width Windows"
author:
- name: Stevie Pederson
  affiliation: 
  - Black Ochre Data Laboratories, Telethon Kids Institute, Adelaide, Australia
  - Dame Roma Mitchell Cancer Researc Laboratories, University of Adelaide
  - John Curtin School of Medical Research, Australian National University
  email: stephen.pederson.au@gmail.com
package: extraChIPs
bibliography: '`r system.file("references.bib", package = "extraChIPs")`'
output:
  github_document
vignette: >
  %\VignetteIndexEntry{Differential Signal Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(
  message = FALSE, warning = FALSE, fig.height = 8, fig.width = 10
)
```

# Introduction

The [GRAVI](https://github.com/smped/GRAVI) workflow, for which this package 
is designed, uses sliding windows for differential signal analysis.
However, the use of fixed-width windows, as is common under DiffBind-style
[@DiffBind2012] approaches is also possible with `extraChIPs`.
This vignette focusses on using conventional peak calls and fixed-width 
approaches to replicate and extend these approaches.

The majority of examples below use heavily reduced datasets to provide
general guidance on using the functions.
Some results may appear trivial as a result, but will hopefully prove far more
useful in a true experimental context.
All data, along with this vignette are available 
[here](https://github.com/smped/extraChIPs_vignette).
Please place all contents of the data directory in a directory named data in 
your own working directory.

# Setup 

## Installation

In order to use the package `extraChIPs` and follow this vignette, we recommend 
using the package `BiocManager` hosted on CRAN.
Once this is installed, the additional packages required for this vignette 
(`tidyverse`, `Rsamtools`, `csaw`, `BiocParallel` and `rtracklayer`) can also 
be installed.

```{r install, eval = FALSE}
if (!"BiocManager" %in% rownames(installed.packages()))
  install.packages("BiocManager")
pkg <- c(
  "tidyverse", "Rsamtools", "csaw", "BiocParallel", "rtracklayer", "edgeR", 
  "patchwork", "extraChIPs", "plyranges", "scales", "here"
)
BiocManager::install(pkg, update = FALSE)
```

Once these packages are installed, we can load them easily

```{r load-packages}
library(tidyverse)
library(Rsamtools)
library(csaw)
library(BiocParallel)
library(rtracklayer)
library(edgeR)
library(patchwork)
library(extraChIPs)
library(plyranges)
library(scales)
library(glue)
library(ggrepel)
library(here)
theme_set(theme_bw())
```


## Data

All data for this vignette is expected to be in a sub-directory of the working 
directory named "data", and all paths will be predicated on this.
Please ensure you have all data in this location, obtained from 
[here](https://github.com/smped/extraChIPs_vignette).

The data itself is ChIP-Seq data targeting the histone mark H3K27ac, and is 
taken from the cell-line MDA-MB-453 under Vehicle and DHT-stimulated conditions.
Using CRCh37 as the reference genome, a subset of regions found on chromosome 10
are included in this dataset for simplicity.

`Seqinfo` objects are the foundation of working with GRanges, so let's define a
suitable object for consistency throughout th analysis.


```{r load-samples}
samples <- here("data", "PRJNA509779.tsv") %>% 
  read_tsv() %>% 
  dplyr::filter(target == "ER") %>% 
  mutate(treatment = factor(treatment, levels = c("E2", "E2DHT")))
accessions <- samples %>% 
  split(f = .$treatment) %>% 
  lapply(pull, "accession")
treat_colours <- c("steelblue", "red3", "grey")
names(treat_colours) <- c(levels(samples$treatment), "Input")
```

```{r bfl}
bfl <- here("data", "ER", glue("{samples$accession}.bam")) %>% 
  BamFileList() %>% 
  setNames(str_remove_all(names(.), ".bam"))
file.exists(path(bfl))
sq <- seqinfo(bfl)
isCircular(sq) <- rep(FALSE, length(seqlevels(sq)))
genome(sq) <- "GRCh37"
```


# Working With Peaks

The provided dataset includes six files produced by `macs2 callpeak` 
[@Zhang2008-ms] in the `narrowPeak` format, and these are able to be easily 
parsed using `extraChIPs`.

```{r load-peaks}
peaks <- here("data", "ER", glue("{samples$accession}_peaks.narrowPeak")) %>% 
  importPeaks(seqinfo = sq)
```

This will import the peaks from all files as a single `GRangesList` object, 
adding the file-name to each element by default.
We can easily modify these names if we so wish.

```{r change-names}
names(peaks) <- str_remove_all(names(peaks), "_peaks.narrowPeak")
```

Once loaded, we can easily check how similar our replicates are.

```{r plot-overlaps, fig.height=6, fig.cap = "*UpSet plot showing overlapping peaks across all replicates*"}
plotOverlaps(
  peaks, min_size = 10, .sort_sets = FALSE, 
  set_col = treat_colours[as.character(samples$treatment)]
)
```

Optionally, specifying a column and a suitable function will produce an 
additional panel summarising that value.
In the following, we'll show the maximum score obtained, highlighting that for 
peaks identified in only one or two replicates, the overall signal intensity is
generally lower.

```{r plot-overlaps-score, fig.cap = "*UpSet plot showing overlapping peaks across all replicates, with the maximum score across all replicates shown in the upper panel.*"}
plotOverlaps(
  peaks, min_size = 10, .sort_sets = FALSE, var = "score", f = "max",
   set_col = treat_colours[as.character(samples$treatment)]
)
```

A common task at this point may be to define consensus peaks within each 
treatment group, by retaining only the peaks found in 2 of the 3 replicates.
The default approach is to take the union of all ranges, with the returned 
object containing logical values for each sample, as well as the number of 
samples where an overlapping peak was found.

If we wish to retain any of the original columns, such as the `macs2 callpeak` 
score, we can simply pass the column names to `makeConsensus()`

```{r consensus-with-score}
consensus_e2 <- makeConsensus(
  peaks[accessions$E2], p = 2/3, var = "score"
  )
consensus_e2dht <- makeConsensus(
  peaks[accessions$E2DHT], p = 2/3, var = "score"
  )
```

Alternatively, we could find the centre of the peaks as part of this process, 
by averaging across the estimated peak centres for each sample.
Whilst this is very common for *transcription factor* peaks, this may be less 
informative for other types of ChIP targets, such as the histone marks we have.

```{r consensus-with-centre}
consensus_e2 <- peaks[accessions$E2] %>% 
  endoapply(mutate, centre = start + peak) %>% 
  makeConsensus(p = 2/3, var = "centre") %>% 
  mutate(centre = vapply(centre, mean, numeric(1)))
consensus_e2dht <- peaks[accessions$E2DHT] %>% 
  endoapply(mutate, centre = start + peak) %>% 
  makeConsensus(p = 2/3, var = "centre") %>% 
  mutate(centre = vapply(centre, mean, numeric(1)))
```

We can also inspect these using `plotOverlaps()` provided we use a `GRangesList`
for the input.

```{r venn-consensus-overlap, fig.width=8, fig.cap = "*Overlap between consensus peaks identified in a treatment-specific manner*"}
GRangesList(
  E2 = granges(consensus_e2), E2DHT = granges(consensus_e2dht)
) %>% 
  plotOverlaps(set_col = treat_colours[names(.)])
```

We could go one step further and define the set of peaks found in either 
treatment.
Given we're being inclusive here, we can leave p = 0 so any peak found in 
either treatment is included.

```{r all_consensus}
greylist <- here("data/chr10_greylist.bed") %>% import.bed(seqinfo = sq)
blacklist <- here("data/chr10_blacklist.bed") %>% import.bed(seqinfo = sq)
union_peaks <- GRangesList(
  E2 = select(consensus_e2, centre), E2DHT = select(consensus_e2dht, centre)
) %>% 
  makeConsensus(var = "centre") %>% 
  mutate(centre = vapply(centre, mean, numeric(1))) %>% 
  filter_by_non_overlaps(c(greylist, blacklist)) 
```


```{r}
w <- 500
centred_peaks <- union_peaks %>% 
  mutate(
    centred = GRanges(
      glue("{seqnames}:{centre - w/2}-{centre + w/2}"), seqinfo = sq
    ),
    union_peak = granges(union_peaks)
  ) %>%
  colToRanges("centred")
```


```{r bfl}
se <- regionCounts(bfl, centred_peaks, ext = 200)
colData(se) <- colData(se) %>% 
  as_tibble(rownames = "accession") %>% 
  left_join(samples) %>% 
  mutate(sample = accession) %>% 
  as.data.frame() %>% 
  column_to_rownames("accession") %>% 
  DataFrame()
assay(se, "logCPM") <- cpm(assay(se, "counts"), lib.size = se$totals, log = TRUE)
```

```{r}
plotAssayDensities(se, assay = "logCPM", colour = "treat") +
  scale_colour_manual(values = treat_colours)
plotAssayRle(se, assay = "logCPM", fill = "treat") +
  scale_fill_manual(values = treat_colours)
plotAssayPCA(se, assay = "logCPM", colour = "treat", label = "sample") +
  scale_colour_manual(values = treat_colours)
```



```{r}
X <- model.matrix(~treatment, data = colData(se))
ls_res <- fitAssayDiff(se, design = X, asRanges = TRUE)
sum(ls_res$FDR < 0.05)
```


```{r}
tmm_res <- fitAssayDiff(se, design = X, norm = "TMM", asRanges = TRUE, fc = 1.2)
sum(tmm_res$FDR < 0.05)
```


```{r}
tmm_res %>% 
  as_tibble() %>% 
  mutate(`FDR < 0.05` = FDR < 0.05) %>% 
  ggplot(aes(logCPM, logFC)) +
  geom_point(aes(colour = `FDR < 0.05`)) +
  geom_smooth(method = "lm", se = FALSE) +
  geom_label_repel(
    aes(label = union_peak), colour = "red",
    data = . %>% dplyr::filter(FDR < 0.05)
  ) +
  scale_colour_manual(values = c("black", "red"))
```


```{r}
gencode <- here("data/gencode.v43lift37.chr10.annotation.gtf.gz") %>% 
  import.gff() %>% 
  filter_by_overlaps(GRanges("chr10:42354900-100000000")) %>% 
  split(.$type)
seqlevels(gencode) <- seqlevels(sq)
seqinfo(gencode) <- sq
```


```{r}
tss <- gencode$transcript %>% 
  resize(width = 1, fix = "start") %>% 
  select(gene_id, ends_with("name")) %>% 
  reduceMC(min.gapwidth = 0)
promoters <- gencode$transcript %>% 
    select(gene_id, ends_with("name")) %>% 
    promoters(upstream = 2500, downstream = 500) %>% 
    reduceMC(simplify = FALSE)
```


```{r}
tmm_mapped_res <- tmm_res %>% 
  colToRanges("union_peak") %>% 
  mapByFeature(genes = gencode$gene, prom = promoters)
```

```{r}
esr1 <- here("data/ER/esr1_chr10.hg19.bed.gz") %>% 
  import.bed(
    colnames = c("chrom", "start", "end", "name", "score"), seqinfo = sq
  ) %>% 
  keepStandardChromosomes()
```

```{r}
tmm_res %>% 
  distanceToNearest(esr1) %>% 
  as_tibble() %>% 
  arrange(distance) %>% 
  mutate(p = seq_along(distance) / nrow(.)) %>% 
  ggplot(aes(distance/1e3, p)) +
  geom_line()
```


# Coverage Plots

```{r}
data("grch37.cytobands")
head(grch37.cytobands)
bwfl <- here::here(
  "data", "ER", glue("{levels(samples$treatment)}_cov_chr10.bw")
) %>% 
  BigWigFileList() %>% 
  setNames(levels(samples$treatment))
```


```{r plot-hfgc-init, fig.cap = "Generic figure from plotHFGC showing coverage for the selected region only"}
gr <- tmm_res %>% 
  arrange(PValue) %>% 
  filter(FDR < 0.05) %>% 
  colToRanges("union_peak") %>% 
  # filter_by_overlaps(tss) %>%
  .[1]
gr <- keepStandardChromosomes(gr)
plotHFGC(gr, coverage = bwfl, cytobands = grch37.cytobands)
```


```{r}
cov_list <- list(ER = bwfl)
cov_colour <- list(ER = treat_colours[levels(samples$treatment)])
gene_models <- gencode$exon %>% 
  select(
    type, gene = gene_id, exon = exon_id, transcript = transcript_id, 
    symbol = gene_name
  ) %>% 
  keepStandardChromosomes()
```

```{r plot-hfgc-with-genes, fig.cap = "*Coverage for our region showing the relationship of signal to annotated genes*"}
plotHFGC(
  gr, cytobands = grch37.cytobands, 
  coverage =  cov_list, linecol = cov_colour,
  genes = gene_models, genecol = "wheat", collapseTranscripts = FALSE,
  zoom = 50, rotation.title = 90, covsize = 10, genesize = 1
)
```

```{r}
cov_list$H3K27ac <- here::here(
  "data", "H3K27ac", glue("{levels(samples$treatment)}_cov_chr10.bw")
) %>% 
  BigWigFileList() %>% 
  setNames(levels(samples$treatment))
cov_colour$H3K27ac <- treat_colours[levels(samples$treatment)]
```

```{r plot-hfgc-with-genes, fig.cap = "*Coverage for our region showing the relationship of signal to annotated genes*"}
feat_list <- list(
  Promoters = GRangesList(Promoter = keepStandardChromosomes(promoters)),
  ESR1 = GRangesList(ESR1 = esr1)
)
plotHFGC(
  gr, cytobands = grch37.cytobands, 
  features = feat_list,
  featcol = list(Promoters = list(Promoter = "yellow2"), ESR1 = list(ESR1 = "royalblue")),
  coverage =  cov_list, linecol = cov_colour,
  genes = gene_models, genecol = "wheat", collapseTranscripts = FALSE,
  zoom = 15, shift = 5e3, rotation.title = 90
)
```

## Profile Heatmaps
